### PR TITLE
Define UnaryOperation, parse BinaryOperation

### PR DESCRIPTION
**Pull request by mernst**

### CODE REVIEW

File: dataflow/src/main/java/org/checkerframework/dataflow/expression/ValueLiteral.java [hunk 1]
```diff
@@ -37,6 +37,58 @@ public ValueLiteral(TypeMirror type, Object value) {
...
+            return Byte.valueOf((byte) -((Byte) value).byteValue());
+        }
+        if (value instanceof Short) {
+            return Short.valueOf((short) -((Short) value).shortValue());
+        }
+        if (value instanceof Integer) {
+            return Integer.valueOf(-((Integer) value).intValue());
+        }
+        if (value instanceof Long) {
+            return Long.valueOf(-((Long) value).longValue());
```
Comment (smillst) [inline dataflow/src/main/java/org/checkerframework/dataflow/expression/ValueLiteral.java:72]: This doesn't work when parsing `-9223372036854775808L`, `Long.MIN_VALUE`.  I added a test case.

File: framework/src/main/java/org/checkerframework/framework/util/JavaExpressionParseUtil.java [hunk 2]
```diff
@@ -628,6 +635,129 @@ public JavaExpression visit(ArrayCreationExpr expr, JavaExpressionContext contex
...
+        }
+
+        @Override
+        public JavaExpression visit(BinaryExpr expr, JavaExpressionContext context) {
+            JavaExpression leftJe = expr.getLeft().accept(this, context);
+            JavaExpression rightJe = expr.getRight().accept(this, context);
+            TypeMirror leftType = leftJe.getType();
+            TypeMirror rightType = rightJe.getType();
+            TypeMirror type;
+            if (types.isSameType(leftType, rightType)) {
```
Comment (smillst) [inline framework/src/main/java/org/checkerframework/framework/util/JavaExpressionParseUtil.java:693]: You can delete this case.  If the types are the same type, then `types.isSubtype` will return true.

File: framework/src/main/java/org/checkerframework/framework/util/JavaExpressionParseUtil.java [hunk 3]
```diff
@@ -628,6 +635,129 @@ public JavaExpression visit(ArrayCreationExpr expr, JavaExpressionContext contex
...
+                    return Tree.Kind.PREFIX_DECREMENT;
+                case PREFIX_INCREMENT:
+                    return Tree.Kind.PREFIX_INCREMENT;
+                default:
+                    throw new Error("unhandled " + op);
+            }
+        }
+
+        @Override
+        public JavaExpression visit(BinaryExpr expr, JavaExpressionContext context) {
```
Comment (smillst) [inline framework/src/main/java/org/checkerframework/framework/util/JavaExpressionParseUtil.java:687]: Would you please add a few test cases for this?  You can do so in `framework/tests/flowexpression/`.

**Review by smillst — APPROVED** [review]

File: framework/src/main/java/org/checkerframework/framework/util/JavaExpressionParseUtil.java (hunk 2 above)
Comment (mernst) [reply framework/src/main/java/org/checkerframework/framework/util/JavaExpressionParseUtil.java:693]: I was thinking that `isSameType` would be more efficient in the common case.  Is that not true?

Comment (smillst) [reply framework/src/main/java/org/checkerframework/framework/util/JavaExpressionParseUtil.java:693]: The implementation of `isSubtype` checks for if the two are equal before the more complicated subtype check.

**Review by smillst — APPROVED** [review]

**Pull request merged**